### ISSUE TITLE
[DEPR]: Devstack

### ISSUE THREAD

**Issue by kdmccormick** [openedx/devstack#907]
### Proposal Date

2022-02-28

### Ticket Acceptance Date

2022-04-01

### Technology Removal Date

2022-08-01

### First Open edX Named Release Without This Functionality

Olive

### Rationale

TODO

### Removal

* The repository https://github.com/openedx/devstack will be archived.
* The various `devstack.py` and `development.py` settings files will be deleted unless otherwise depended upon.
* The various `docker-compose.yml` files and Makefile targets that some IDAs (eg enterprise-catalog) use in order to integrate with Devstack.
* 2U-specific stuff:
  * The Jenkins jobs that build `edxops/*` Docker images.
  * The `edxops` DockerHub organization. 

### Replacement

Tutor will replace Devstack.
* Specifically, [Tutor Nightly](https://docs.tutor.overhang.io/tutorials/nightly.html) will be the closest replacement, as it tracks runs master branches just like Devstack did.
* Several Tutor plugins will be needed in order for Tutor to fully replace devstack. Many already exist, such as [tutor-discovery](https://github.com/overhangio/tutor-discovery/). 

The [Tutor Adoption Initiative](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3315335223/Tutor+Adoption+Initiative) will smooth the transition between Devstack and Tutor.

### Deprecation

TODO

### Migration

TODO

### Additional Info

TODO

**Comment by kdmccormick** [comment]
Relates to https://github.com/openedx/build-test-release-wg/issues/138

**Comment by dianakhuang** [comment]
Since this is primarily a 2U project right now, we are planning on announcing this and gathering feedback. 

2U is still planning on using this for the near to medium term until we're confident we can migrate successfully onto Tutor, but we can move it out of the openedx GitHub org to minimize confusion.

**Comment by kdmccormick** [comment]
In general, sounds great!

Only thing is...  MFE development currently sucks in Tutor >.< So nearly all frontend devs still use Devstack when they're trying to change MFEs.

Brian Smith is [actively working on rectifying this](https://github.com/openedx/wg-developer-experience/issues/123), with promising results. Would you be willing to delay the deprecation further until Tutor has MFE dev sorted out?

**Comment by dianakhuang** [comment]
Yup! I think we're still trying to understand the requirements needed here. I'll push out the acceptance date for another few months.

**Comment by kdmccormick** [comment]
@dianakhuang So, I know I had originally been a proponent of deprecating both configuration and devstack "into the edx org", under the understanding that if only 2U is using some parts of the project, then it's easier for both sides if we transfer those parts over to 2U. Since then, I've talked about this with my team at Axim, and come to understand that the situation is more nuanced than that.

The code in the openedx org is built from a mix of Axim-owned contributions and community-contributor-owned contributions (licensed to Axim under the CLA). We hold the code in this org for the good of all current & former Open edX contributors and users, and as the project's stewards we have a mandate to maintain ownership and ensure availability of all its repositories, even unsupported repositories.

Our procedure of archiving repos by transferring to the openedx-unsupported org lets us mark a repo as "unsupported" while still guaranteeing the repo's availability and keeping its ownership clear. So, if/when devstack and configuration are deprecated, we'll need to move them there.

**Comment by dianakhuang** [comment]
@kdmccormick that sounds reasonable! 2U/edX.org should be able to then clone the repositories for our own use case after they are put into unsupported, then?

Obviously, this is our first stab and understanding how this process is going to work going forward.

**Comment by kdmccormick** [comment]
Yes, you could fork from openedx-supported after that full deprecation & archival process.

Even in that case, though, I'd be curious what we could do to help keep 2U involved with community tooling.

Would you be forking devstack & configuration as a temporary stop-gap until you're able to transition off of them, or with the intent of using them indefinitely?

* If you're using them as a stop-gap, then I bet some other community members are also using them as a stop-gap too. It could be worth keeping these repos un-archived until the replacement options (Tutor or otherwise) are viable to all or almost all community members.
* If you plan to use them indefinitely, then I bet some some other community members would see value in using them indefinitely, too, which I would see as an argument for keeping the repos un-archived.

> Obviously, this is our first stab and understanding how this process is going to work going forward.

Same here :) learning as we go!

**Comment by dianakhuang** [comment]
We are going to block acceptance because devstack is still used heavily by people who need to do development on MFEs outside of 2U.

@arbrandes @kdmccormick could you add more context on this?

**Comment by kdmccormick** [comment]
@arbrandes , I know you mentioned that RaccoonGang are heavy users of Devstack. I wonder if they'd be willing to take on maintenance. If not, and if nobody else is willing to, then we should probably let this DEPR move forward and help MFE developers move towards Tutor.

**Comment by feanil** [comment]
@dianakhuang given recent changes at 2U, is 2U still willing to maintain this repository?  Or should we complete the deprecation and remove this from the `openedx` org?

**Comment by dianakhuang** [comment]
@feanil 2U is still willing to maintain this repository for now. We would like to keep in in the `openedx` org for now if that's okay with everyone else.

**Comment by feanil** [comment]
I think this is fine for now but since we're trying to move away from devstack for the community, at some point it will make senses to move it out.  In the meantime, I'd like to mark this ticket and the repo as deprecated and make sure the relevant messaging exists in the README for this.  What do you think?

**Comment by dianakhuang** [comment]
Yup, let's add some text to the README and move this ticket to the deprecated column.

**Comment by feanil** [comment]
Cool, is that something you can take on as maintainers?

**Comment by dianakhuang** [comment]
Yup, I'll try to get on it today and spread the info around 2U so no one freaks out.

**Comment by feanil** [comment]
We should move this back to Proposed and re-communicate that 2U wants to stop maintaining this as a part of Open edX.  Either someone else needs to maintain this for the community or we can archive and 2U and any other interested parties can fork it.
